[FIXED JENKINS-27441] avoid IOOBE in getLog when called with 0

vb-linetco opened the pull request:
return an empty list when called with maxLines = 0 instead

olivergondza commented:
+1, can you add a unit test?

olivergondza commented:
Unittest does not reproduce the problem:
```
getLogReturnsAnEmptyListWhenCalledWith0(hudson.model.RunTest)  Time elapsed: 0.449 sec  <<< ERROR!
java.io.FileNotFoundException: .../jenkins/core/target/0/log (No such file or directory)
	at java.io.FileInputStream.open(Native Method)
	at java.io.FileInputStream.<init>(FileInputStream.java:146)
	at hudson.model.Run.getLog(Run.java:1942)
	at hudson.model.RunTest.getLogReturnsAnEmptyListWhenCalledWith0(RunTest.java:128)
```

When I create the log file before trying to read it, it succeeds. What is the complete stacktrace of the problem you are fixing? Also, consider creating JIRA isues with all the detail for bugs.

vb-linetco commented:
I've created an issue in JIRA, i think you have to add at least one line to the log to trigger the exception.
I've no clue how to add some log lines in the test case. Any pointers how I can add some log entries would be appreciated.

olivergondza commented:
> Any pointers how I can add some log entries would be appreciated.

IIRC `org.apache.commons.io.FileUtils` provides such helper functions.

daniel-beck commented:
I'd prefer a squashed PR. 9 commits for this change is a bit much IMO.

Renaming PR so a merge will automatically close the issue (https://wiki.jenkins-ci.org/display/JENKINS/contributing#contributing-CommitMessages)…

vb-linetco committed:
avoid IndexOutOfBoundsException when called with 0
return an empty list when called with maxLines = 0
added unit test for getLog(0) returning an empty list

bootstraponline commented:
I ran into this issue today on cloudbees. It'd be nice to get this fixed.

daniel-beck commented:
I missed that the PR was updated to use a single commit. Sorry about that!

Cannot be merged cleanly anymore, so please update the PR.

vb-linetco commented:
Updated the PR, can be automatically merged again.

bootstraponline commented:
I think you need to do:

```
git pull --rebase git@github.com:jenkinsci/jenkins.git
git push -f
```

This PR should show exactly one commit, not 43.

vb-linetco commented:
Thx for hint, forgot to rebase. Should be one commit now.

bootstraponline commented:
Looks great. Thanks!

oleg-nenashev commented:
LGTM

jtnord commented on core/src/test/java/hudson/model/RunTest.java:
```diff
@@ -128,4 +131,18 @@ public void getDurationString() throws IOException {
+        w.close();
+        List<String> logLines = r.getLog(0);
+        assertTrue(logLines.isEmpty());
+    }
```
the file is never deleted AFAICT and will litter the temp directory.

hrzlgnm replied:
should not be littering anymore

vb-linetco replied:
should not be littering in temp directory anymore

jtnord commented:
:+1:

oleg-nenashev commented:
@vb-linetco 
Merged, thanks for the fix!
